### PR TITLE
feat: add historical statistics endpoint

### LINKED ISSUE #37: Endpoint to retrieve monthly propotion of active tracked services
Goal is to get this kind of presentation

image

I believe @vviers, you were the one who produced this image if I recall correctly.

So I think the endpoint could be

- `GET /graph_services/v1` (to respect the naming `<action>_<item>` that seems to be used)
```
{
graph: [ 
  {
    date:"2021-08-31",
    nbTrackedServices: 200,
    nbActiveServices: 100,
  },
  {
     ...
  }
]
}
```
Let me know what you think.

### PR DESCRIPTION
**Pull request by vviers**
Closes #37 

Let me know what you think :)

### CODE REVIEW

**Comment by martinratinaud** [comment]
Thanks @vviers 

I tried to launch it using

```
python -m pip install -r requirements.txt
cd app
export PYTHONPATH=.
uvicorn main:app --reload
```

But I get the following error

```
ERROR:    Traceback (most recent call last):
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/site-packages/starlette/routing.py", line 526, in lifespan
    async for item in self.lifespan_context(app):
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/site-packages/starlette/routing.py", line 467, in default_lifespan
    await self.startup()
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/site-packages/starlette/routing.py", line 502, in startup
    await handler()
  File "./main.py", line 57, in startup_event
    logger.info(f"Dataset version : {read_dataset()}")
  File "./main.py", line 48, in read_dataset
    return dataset_file.read_text().strip("\n")
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/pathlib.py", line 1206, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/pathlib.py", line 1193, in open
    opener=self._opener)
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/pathlib.py", line 1046, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: 'latest_dataset.txt'
```

Any idea why?

**Comment by vviers** [comment]
> Any idea why?

Yes ! The issue is that you need to have the dataset and the `latest_dataset.txt` files inside the `app/` directory (cf. the paths specified in`app/config.py`)

In Docker, everything is set up nicely, but if you want to run the API directly in python you might need to tweak a few things.

**Comment by vviers** [comment]
To make it work in full python you need to do :

```

./download_dataset.sh
python -m pip install -r requirements.txt
export PYTHONPATH=".":"./app/"
uvicorn main:app --reload
```

I'll update the README :)

File: README.md [hunk 1]
```diff
@@ -20,19 +20,21 @@ The API is served at `localhost`.
...
-You will also need a local copy of the dataset. By default, it is assumed to be in `./app/dataset` but you can always change this by editing `./app/config.py`.
+You will also need a local copy of the dataset. By default, it is assumed to be in `./dataset` but you can always change this by editing `./app/config.py`.
+
+To download the dataset, run :
+`./download_dataset.sh`
 
 Install requirements
 
 ```sh
 python -m pip install -r requirements.txt
```
Comment (martinratinaud) [inline README.md:31]: @vviers I see a requirements-dev.txt but do not see it in the docs

Can you please add it?

File: README.md [hunk 2]
```diff
@@ -20,19 +20,21 @@ The API is served at `localhost`.
 
 This API was built using `python3.8`. We suggest you use a virtual environment.
 
-You will also need a local copy of the dataset. By default, it is assumed to be in `./app/dataset` but you can always change this by editing `./app/config.py`.
+You will also need a local copy of the dataset. By default, it is assumed to be in `./dataset` but you can always change this by editing `./app/config.py`.
+
+To download the dataset, run :
+`./download_dataset.sh`
```
Comment (martinratinaud) [inline README.md:26]: download_dataset.sh seems to not be executable

Should launch `chmod a+x download_dataset.sh`

File: app/dataset_parser.py [hunk 3]
```diff
@@ -48,6 +48,16 @@ def list_all_services_doc_types(self, multiple_versions_only: bool = False) -> l
             dict_out[service].append(doc_type)
         return dict(dict_out)
 
+    def get_stats(self):
+        """
+        Extract basic info for every CGU in historical dataset
```
Comment (martinratinaud) [inline app/dataset_parser.py:53]: let's use document instead of CGU as it is not the name of the product anymore

File: app/dataset_parser.py [hunk 4]
```diff
@@ -175,3 +185,42 @@ def run(self):
 
     def to_dict(self):
         return self.output
+
+
+class CGU:  # pylint: disable=too-few-public-methods
+    """
+    A CGU object.
```
Comment (martinratinaud) [inline app/dataset_parser.py:192]: same

File: app/main.py [hunk 5]
```diff
@@ -189,3 +190,34 @@ async def get_version_at_date(
...
+    Returns a JSON object with monthly statistics about tracked services.
+    """
+    dataset = CGUsDataset(Path(CGUS_DATASET_PATH))
+    stats = dataset.get_stats()
+    data = pd.DataFrame.from_dict(stats, orient="index")
+    data.sort_values(by="date", inplace=True)
+    data["year_month"] = data.date.dt.to_period("M")
+    over_years = data.groupby("year_month", as_index=False).agg(
+        services_tracked=("service", "unique"),
+        n_services_tracked=("service", pd.Series.nunique),
```
Comment (martinratinaud) [inline app/main.py]: I would be in favor of having on homogeneous naming

n_services_tracked
n_services_unique

Also I find n_services_tracked misleading as we are indeed tracking all services but only few are active

So I would go for
- n_services_tracked -> n_services_active
- n_unique_services -> n_services_tracked

What do you think?

**Review by martinratinaud — CHANGES_REQUESTED** [review]
Apart from those remarks, I tested and it seems to be good, thanks

File: app/dataset_parser.py (hunk 4 above)
Comment (vviers) [reply app/dataset_parser.py:192]: I agree that the name of the project has changed, but to me the object represented here is still a CGU
I'm happy with replacing all occurences of CGU in the codebase with "Document" but I feel like the code could be a little confusing... What do you think ?

File: README.md (hunk 1 above)
Comment (vviers) [reply README.md:31]: It is mentioned in the README a little below

File: README.md (hunk 2 above)
Comment (vviers) [reply README.md:26]: done

File: app/main.py (hunk 5 above)
Comment (vviers) [reply app/main.py]: done

File: app/dataset_parser.py (hunk 4 above)
Comment (martinratinaud) [reply app/dataset_parser.py:192]: I'm ok, let's leave it that way
it should be ContractualDocument more than Document anyway

**Review by martinratinaud — APPROVED** [review]

**Pull request merged**